File: 240202/challenge121.c
#include <wiringPi.h>
#include <stdio.h>
#include <unistd.h>

#define LED_PIN 4      // 라즈베리 파이의 4번 핀을 사용
#define BUTTON_PIN1 22

int ledStatus = 0;

void buttonCallback(void) {
    if (ledStatus) {
        // LED가 켜져 있으면 끄기
        digitalWrite(LED_PIN, LOW);
        ledStatus = 0;
    } else {
        // LED가 꺼져 있으면 켜기
        digitalWrite(LED_PIN, HIGH);
        ledStatus = 1;
    }
}

int main(void) {
    if (wiringPiSetup() == -1) {
        printf("wiringPi setup failed\n");
        return 1;
    }

    pinMode(BUTTON_PIN1, INPUT);
    pullUpDnControl(BUTTON_PIN1, PUD_UP);

    pinMode(LED_PIN, OUTPUT);

    wiringPiISR(BUTTON_PIN1, INT_EDGE_FALLING, &buttonCallback);

    while (1) {
        // 프로그램이 계속 실행되도록 유지
        usleep(100000);  // 0.1초 대기
    }

    return 0;
}

File: 240202/challenge121.py
import RPi.GPIO as GPIO
import time

# GPIO 핀 번호 설정
led_pin = 4  # 라즈베리 파이의 4번 핀을 사용
button_pin1 = 22

# GPIO 설정
GPIO.setmode(GPIO.BCM)
GPIO.setup(button_pin1, GPIO.IN, pull_up_down=GPIO.PUD_UP)

GPIO.setup(led_pin, GPIO.OUT)
GPIO.setup(button_pin1, GPIO.IN)

led_status = False

def button_callback(channel):
    global led_status
    if led_status:
        # LED가 켜져 있으면 끄기
        GPIO.output(led_pin, GPIO.LOW)
        led_status = False
    else:
        # LED가 꺼져 있으면 켜기
        GPIO.output(led_pin, GPIO.HIGH)
        led_status = True

GPIO.add_event_detect(button_pin1, GPIO.FALLING, callback=button_callback, bouncetime=300)

try:
    while True:
        # 프로그램이 계속 실행되도록 유지
        time.sleep(0.1)
        
except KeyboardInterrupt:
    GPIO.cleanup()
